File: source/data/mappers/src/jvmMain/kotlin/com/voxeldev/tgdrive/mappers/user/UserMapper.kt
package com.voxeldev.tgdrive.mappers.user

import com.voxeldev.tgdrive.main.api.profile.Profile
import com.voxeldev.tgdrive.mappers.media.FileMapper
import org.drinkless.tdlib.TdApi.ProfilePhoto
import org.drinkless.tdlib.TdApi.User
import com.voxeldev.tgdrive.main.api.profile.ProfilePhoto as CommonProfilePhoto

/**
 * @author nvoxel
 */
class UserMapper(
    private val fileMapper: FileMapper,
) {

    fun toResponse(user: User): Profile =
        Profile(
            id = user.id,
            firstName = user.firstName,
            lastName = user.lastName,
            username = user.usernames.editableUsername,
            phoneNumber = user.phoneNumber,
            profilePhoto = user.profilePhoto?.let { user.profilePhoto.toResponse() },
            isContact = user.isContact,
            isMutualContact = user.isMutualContact,
            isVerified = user.isVerified,
            isSupport = user.isSupport,
            restrictionReason = user.restrictionReason,
            isScam = user.isScam,
            isFake = user.isFake,
            haveAccess = user.haveAccess,
            languageCode = user.languageCode,
        )

    private fun ProfilePhoto.toResponse(): CommonProfilePhoto =
        CommonProfilePhoto(
            id = id,
            big = fileMapper.toResponse(big),
            small = fileMapper.toResponse(small),
        )
}

File: iosApp/iosApp/data/mappers/user/UserMapper.swift
//
//  UserMapper.swift
//  iosApp
//
//  Created by nVoxel on 25.04.2024.
//  Copyright © 2024 orgName. All rights reserved.
//

import Foundation
import TDLibKit
import TGDriveKit

class UserMapper {
    
    private let fileMapper: FileMapper
    
    init(fileMapper: FileMapper) {
        self.fileMapper = fileMapper
    }
    
    func toResponse(user: User) -> CommonProfile {
        return CommonProfile(
            id: user.id,
            firstName: user.firstName,
            lastName: user.lastName,
            username: user.usernames!.editableUsername,
            phoneNumber: user.phoneNumber,
            profilePhoto: user.profilePhoto != nil ? toReponse(profilePhoto: user.profilePhoto!) : nil,
            isContact: user.isContact,
            isMutualContact: user.isMutualContact,
            isVerified: user.isVerified,
            isSupport: user.isSupport,
            restrictionReason: user.restrictionReason,
            isScam: user.isScam,
            isFake: user.isFake,
            haveAccess: user.haveAccess,
            languageCode: user.languageCode
        )
    }
    
    private func toReponse(profilePhoto: ProfilePhoto) -> CommonProfilePhoto {
        return CommonProfilePhoto(
            id: profilePhoto.id.rawValue,
            small: fileMapper.toResponse(file: profilePhoto.small),
            big: fileMapper.toResponse(file: profilePhoto.big)
        )
    }
    
}
